examples/run_summarization_finetuning.py

```diff
@@ -0,0 +1,625 @@
+                    except IndexError:  # skip ill-formed stories
+                        continue
+
+                story_seq = _fit_to_block_size(story_ids, block_size)
+                self.source_examples.append(story_seq)
+
+                summary_seq = _fit_to_block_size(summary_ids, block_size)
+                self.target_examples.append(summary_seq)
+
+        logger.info("Saving features into cache file %s", cached_features_file)
+        with open(cached_features_file, "wb") as sink:
+            pickle.dump(
+                {"source": self.source_examples, "target": self.target_examples},
+                sink,
+                protocol=pickle.HIGHEST_PROTOCOL,
+            )
+
+    def __len__(self):
+        return len(self.source_examples)
+
+    def __getitem__(self, items):
+        return (
+            torch.tensor(self.source_examples[items]),
+            torch.tensor(self.target_examples[items]),
+        )
+
+
+def process_story(raw_story, tokenizer):
+    """ Extract the story and summary from a story file.
+
+    Attributes:
+        raw_story (str): content of the story file as an utf-8 encoded string.
+
+    Raises:
+        IndexError: If the stoy is empty or contains no highlights.
+    """
+    file_lines = list(
+        filter(lambda x: len(x) != 0, [line.strip() for line in raw_story.split("\n")])
+    )
+
+    # for some unknown reason some lines miss a period, add it
+    file_lines = [_add_missing_period(line) for line in file_lines]
+
+    # gather article lines
+    story_lines = []
+    lines = deque(file_lines)
+    while True:
+        try:
+            element = lines.popleft()
+            if element.startswith("@highlight"):
+                break
+            story_lines.append(element)
+        except IndexError as ie:  # if "@highlight" absent from file
+            raise ie
+
+    # gather summary lines
+    highlights_lines = list(filter(lambda t: not t.startswith("@highlight"), lines))
+
+    # join the lines. This sequence is specific to Lapata & Liu (2019).
+    # Edit when you are adding a new model.
+    story_ids = [
+        tokenizer.add_special_tokens_single_sequence(tokenizer.encode(story))
+        for story in story_lines
+    ]
+    summary_ids = [
+        tokenizer.add_special_tokens_single_sequence(tokenizer.encode(highlight))
+        for highlight in highlights_lines
+    ]
+
+    flattened_story_ids = [token for sentence in story_ids for token in sentence]
+    flattened_summary_ids = [token for sentence in summary_ids for token in sentence]
+
+    return flattened_story_ids, flattened_summary_ids
+
+
+def _add_missing_period(line):
+    END_TOKENS = [".", "!", "?", "...", "'", "`", '"', u"\u2019", u"\u2019", ")"]
+    if line.startswith("@highlight"):
+        return line
+    if line[-1] in END_TOKENS:
+        return line
+    return line + "."
+
+
+def _fit_to_block_size(sequence, block_size):
```

Comment:
Question: what do we do when we brutally cut mid-sentence? Remove beginning of the sentence and pad? Leave it as is?

Reply:
leave it as is I would say

Reply:
:+1: